utils/nvme.go

```diff
@@ -416,10 +417,178 @@ func (n *Nvme) Format(ctx context.Context, device string, ses SecureEraseSetting
+		return 0, err
+	}
+
+	// parse namespace id from stdout which looks like: `create-ns: Success, created nsid:1`
+	out := bytes.TrimSpace(result.Stdout)
+	parts := bytes.Split(out, []byte(":"))
+	if len(parts) != 3 {
+		return 0, fmt.Errorf("unable to parse nsid: %w", io.ErrUnexpectedEOF)
+	}
+	nsid, err := strconv.Atoi(string(parts[2]))
+	return uint(nsid), err
+}
+
+func (n *Nvme) deleteNS(ctx context.Context, device string, namespaceID uint) error {
+	nsid := strconv.Itoa(int(namespaceID))
+	n.Executor.SetArgs("delete-ns", device, "--namespace-id="+nsid)
+	_, err := n.Executor.Exec(ctx)
+	return err
+}
+
+func (n *Nvme) attachNS(ctx context.Context, device string, controllerID, namespaceID uint) error {
+	cntlid := strconv.Itoa(int(controllerID))
+	nsid := strconv.Itoa(int(namespaceID))
+	n.Executor.SetArgs("attach-ns", device, "--controllers="+cntlid, "--namespace-id="+nsid)
+	_, err := n.Executor.Exec(ctx)
+	return err
+}
+
+func (n *Nvme) idNS(ctx context.Context, device string, namespaceID uint) ([]byte, error) {
+	nsid := strconv.Itoa(int(namespaceID))
+	n.Executor.SetArgs("id-ns", "--output-format=json", device, "--namespace-id="+nsid)
+	result, err := n.Executor.Exec(ctx)
+	return result.Stdout, err
+}
+
+func (n *Nvme) ResetNS(ctx context.Context, device string) error { // nolint:gocyclo
+	out, err := n.cmdListCapabilities(ctx, device)
+	if err != nil {
+		return err
+	}
+
+	ctrl := struct {
+		CNTLID  uint `json:"cntlid"`
+		TNVMCAP uint `json:"tnvmcap"`
+	}{}
+	err = json.Unmarshal(out, &ctrl)
+	if err != nil {
+		return err
+	}
+
+	namespaces, err := n.listNS(ctx, device)
+	if err != nil {
+		return err
+	}
+
+	// we need to have at least 1 namespace so we can interogate the features supported
+	if len(namespaces) == 0 {
+		var nsid uint
+		nsid, err = n.createNS(ctx, device, ctrl.TNVMCAP/512, 0)
+		if err != nil {
+			return err
+		}
+
+		err = n.attachNS(ctx, device, ctrl.CNTLID, nsid)
+		if err != nil {
+			return err
+		}
+
+		namespaces, err = n.listNS(ctx, device)
+		if err != nil {
+			return err
+		}
+		if len(namespaces) == 0 {
+			err = fmt.Errorf("%s: failed to find namespaces: %w", device, io.ErrUnexpectedEOF)
+			return err
+		}
+	}
+
+	out, err = n.idNS(ctx, device, namespaces[0])
+	if err != nil {
+		return err
+	}
+	ns := struct {
+		LBAFS []struct {
+			DS int `json:"ds"`
+			MS int `json:"ms"`
+			RP int `json:"rp"`
+		} `json:"lbafs"`
+	}{}
+	err = json.Unmarshal(out, &ns)
+	if err != nil {
+		return err
+	}
+
+	ds := 0
+	var flbas uint8
+	for i, lbafs := range ns.LBAFS {
+		if lbafs.DS > ds {
+			ds = lbafs.DS
+			flbas = uint8(i)
+		}
+	}
+
+	// info gathered and looks ok, lets get dangerous
+
+	// delete all namespaces
+	for _, ns := range namespaces {
+		err = n.deleteNS(ctx, device, ns)
+		if err != nil {
+			return err
+		}
+	}
+
+	// figure out nsze and ncap in terms of blocksize, we want both to be the same
+	var nsid uint
+	nsid, err = n.createNS(ctx, device, ctrl.TNVMCAP/(1<<ds), flbas)
```

Comment:
Thanks for calling this out, I've reworked it so this function takes in the full size and wanted block size and does the math internally. I've also added a comment for the bit shifting (I moved it into the loop instead of here).

Reply:
Thank you!!!